src/redis_lib/mod.rs

```diff
@@ -10,44 +11,134 @@ use tracing::error;
+    get_email_stats(con, &hash_email(email)).await.0
+}
+
+/// gets how often an email was used to verify an account (not necessarily successful)
+#[allow(unused)]
+pub async fn get_email_used_verification(con: &mut Connection, email: &str) -> u64 {
+    get_email_stats(con, &hash_email(email)).await.1
+}
+
+async fn set_email_stats(con: &mut Connection, email_hash: &str, successful: u64, used: u64) {
+    let command: Result<(), _> = con.set(email_hash, format!("{successful},{used}")).await;
+    if let Err(err) = command {
+        error!(
+            error = err.to_string(),
+            "Could not set stats for email {email_hash}, ({successful},{used})"
+        );
     }
-    if let Err(err) = client.get_connection() {
-        error!(error=err.to_string(), "Failed connecting to Redis");
-        return None;
+}
+
+async fn get_email_stats(con: &mut Connection, email_hash: &str) -> (u64, u64) {
+    let command: Result<Option<String>, _> = con.get(email_hash).await;
+    if let Err(err) = command {
+        error!(error = err.to_string(), "Could not read email stats");
+        return (0, 0);
+    }
+    let redis_value = command.unwrap();
+
+    if redis_value.is_none() {
+        return (0, 0);
+    }
+    let redis_value = redis_value.unwrap();
+
+    let values: Vec<&str> = redis_value.split(',').collect();
+    if values.len() != 2 {
+        error!("Redis returned invalid value for {email_hash}")
+    }
+    let used_success = values[0].parse();
+    if used_success.is_err() {
+        error!("Failed to parse used success value for {email_hash}");
+        return (0, 0);
     }
-    return Some(client);
+
+    let used = values[1].parse();
+    if used.is_err() {
+        error!("Failed to parse used value for {email_hash}");
+        return (0, 0);
+    }
+    (used_success.unwrap(), used.unwrap())
 }
 
-/// Opens a connection to the redis database
-pub fn connect(client: Client) -> Option<Connection> {
-    match client.get_connection() {
-        Ok(val) => Some(val),
-        Err(err) => {
-            error!(error=err.to_string(), "Failed connecting to Redis");
-            None
-        }
+fn hash_email(email: &str) -> String {
+    let mut hasher = sha1::Sha1::new();
+    hasher.update(email);
+    hex::encode(hasher.finalize())
+}
+
+/// The key under which redis stores the main discord channel to put logs into
+const MAIN_DISCORD_LOG_CHANNEL_KEY: &str = "main_discord_log_channel";
+
+#[allow(unused)]
+pub async fn set_main_discord_log_channel(con: &mut Connection, channel_id: &str) {
+    let command: Result<(), _> = con.set(MAIN_DISCORD_LOG_CHANNEL_KEY, channel_id).await;
+    if let Err(err) = command {
+        error!(
+            error = err.to_string(),
+            "Could not set discord main log channel"
+        );
+    }
+}
+
+#[allow(unused)]
+pub async fn get_main_discord_log_channel(con: &mut Connection) -> Option<String> {
+    con.get(MAIN_DISCORD_LOG_CHANNEL_KEY).await.ok()
+}
```

Comment:
Change type of function so that it is clear if there is a problem with the redis connection or that there is no value set.
```suggestion
pub async fn get_main_discord_log_channel(con: &mut Connection) -> Result<Option<String>, ()> {
    let command: Result<Option<String>, _> = con.get(MAIN_DISCORD_LOG_CHANNEL_KEY).await;
    command.or_else(|err| {
        error!(error = err.to_string(), "Could not read email stats");
        Err(())
    })
}
```